src/java/src/main/java/com/dnanexus/DXSearch.java

```diff
@@ -989,14 +1022,110 @@ private T getDataObjectInstanceFromResult(FindDataObjectsResponse.Entry e) {
+
+            int nextElementIndex;
+
+            private ResultIterator() {
+                this.nextElementIndex = 0;
+            }
+
+            @Override
+            public boolean hasNext() {
+                return nextElementIndex < response.results.size();
+            }
+
+            @Override
+            public T next() {
+                return getDataObjectInstanceFromResult(response.results.get(nextElementIndex++));
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException();
+            }
+
+            @SuppressWarnings("unchecked")
+            private T getDataObjectInstanceFromResult(FindDataObjectsResponse.Entry e) {
+                DXDataObject dataObject = null;
+                DXContainer container = DXContainer.getInstance(e.project);
+                if (e.describe != null) {
+                    dataObject =
+                            DXDataObject.getInstanceWithCachedDescribe(e.id, container, env,
+                                    e.describe);
+                } else {
+                    dataObject = DXDataObject.getInstanceWithEnvironment(e.id, container, env);
+                }
+
+                if (classConstraint != null) {
+                    if (!dataObject.getId().startsWith(classConstraint + "-")) {
+                        throw new IllegalStateException("Expected all results to be of type "
+                                + classConstraint + " but received an object with ID "
+                                + dataObject.getId());
+                    }
+                }
+                // This is an unchecked cast, but the callers of this class
+                // should have set T appropriately so that it agrees with
+                // the class constraint (if any). If something goes wrong
+                // here, either that code is incorrect or the API server
+                // has returned incorrect results.
+                return (T) dataObject;
+            }
+        }
+
+        /**
+         * Initializes result set page object with the specified page size.
+         */
+        private SearchPage(FindDataObjectsRequest request, String classConstraint,
+                                      DXEnvironment env) {
+            this.classConstraint = classConstraint;
+            this.env = env;
+            this.response = DXAPI.systemFindDataObjects(request, FindDataObjectsResponse.class, env);
+        }
+
+        /**
+         * Returns an amount of items on findDataObjects results page
+         */
+        public int size() {
+            return response.results.size();
+        }
+
+        @Override
+        public Iterator<T> iterator()
+        {
+            return new ResultIterator();
+        }
+
+        public boolean hasNext() {
+            return response.next != null && !response.next.isNull();
+        }
+
+        public JsonNode getNext() {
+            return response.next;
```

Comment:
Since we're returning an object to the caller here we should make a defensive copy. Maybe JsonNode.deepCopy, or I don't know if there's a way to present an immutable view of a JsonNode?